dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/relation/RelationModelManager.java

```diff
@@ -0,0 +1,623 @@
+
+        for (Set<DependencyGraph<Relation>.Node> component : dependencyGraph.getSCCs()) {
+            for (DependencyGraph<Relation>.Node node : component) {
+                Relation r = node.getContent();
+                if (r.isRecursive()) {
+                    RecursiveGraph recGraph = new RecursiveGraph();
+                    recGraph.setName(r.getNameOrTerm() + "_rec");
+                    relGraphs.add(recGraph);
+                    relGraphCache.put(r, recGraph);
+                }
+            }
+            for (DependencyGraph<Relation>.Node node : component) {
+                Relation r = node.getContent();
+                RelationGraph rg = createGraph(r);
+                if (r.isRecursive()) {
+                    RecursiveGraph recGraph = (RecursiveGraph) relGraphCache.get(r);
+                    recGraph.setConcreteGraph(rg);
+                } else {
+                    relGraphs.add(rg);
+                }
+            }
+        }
+
+        Set<CAATPredicate> predicates = new HashSet<>(relGraphs);
+        PredicateHierarchy hierarchy = new PredicateHierarchy(predicates);
+        hierarchy.initializeToDomain(new EventDomain(oldModel));
+
+        // Populate graphs of base relations.
+        for (CAATPredicate basePred : hierarchy.getBasePredicates()) {
+            if (basePred.getClass() == StaticWMMSet.class
+                || basePred.getClass() == StaticDefaultWMMGraph.class
+            ) { continue; }
+            Relation r = relGraphCache.inverse().get((RelationGraph) basePred);
+            r.getDefinition().accept(graphPopulator);
+        }
+
+        // Do the computation.
+        hierarchy.populate();
+
+        for (CAATPredicate pred : hierarchy.getPredicateList()) {
+            if (pred.getClass() == StaticWMMSet.class) { continue; }
+            Relation r = relGraphCache.inverse().get((RelationGraph) pred);
+            if (relModelCache.containsKey(r)) {
+                RelationModel rm = relModelCache.get(r);
+                ((RelationGraph) pred).edgeStream()
+                    .forEach(e -> rm.addEdgeModel(getOrCreateEdgeModel(e)));
+           }
+        }
+
+        for (Relation r : relsToExtract) {
+            executionModel.addRelation(r, relModelCache.get(r));
+        }
+    }
+
+    private void createModel(Relation r, String name) {
+        if (relModelCache.containsKey(r)) {
+            relModelCache.get(r).addName(name);
+        } else {
+            relModelCache.put(r, new RelationModel(r, name));
+        }
+    }
+
+    private RelationGraph createGraph(Relation r) {
+        RelationGraph rg;
+        try {
+            rg = r.getDefinition().accept(graphBuilder);
+        } catch (UnsupportedOperationException e) {
+            rg = new StaticDefaultWMMGraph(r, getContextWithFullWmm().getAnalysisContext()
+                                                                     .requires(RelationAnalysis.class)
+            );
```

Comment:
What is the fallback for? This looks very dangerous

Reply:
The visitor `graphBuilder` creates `SimpleGraph` for base relations and the specific graph for derived ones, e.g. `InverseGraph` for `Inverse`. But for some relations that don't have the specific graph, like `DomainIdentity`, we create the static default graph for them.

Reply:
The fallback is wrong and will generate wrong graphs. But you are indeed right that `DomainIdentity` is missing on the CAAT side and this is a bug I believe. I'm surprised that this has not caused any issues so far.
I will fix it.

Reply:
Which graph should be created for GPU relations then?

Reply:
You can use `DynamicDefaultWMMGraph` which should work as long as the relation gets encoded(*). In #785 I made sure that those get encoded when doing refinement.


(*) We do not necessarily encode relations fully: see https://github.com/hernanponcedeleon/Dat3M/pull/742#issuecomment-2379226002